### ISSUE TITLE
Should the addon and website always use the same language?

### ISSUE THREAD

**Issue by 6a68** [mozilla-services/pageshot#2374]
Based on playing around with [this localized sample addon](https://github.com/mdn/webextensions-examples/tree/master/notify-link-clicks-i18n/) and a German copy of Firefox, it looks like the WebExtension i18n system uses the UI language (the language of the browser), and not the content language (language picked by the user in the FF settings menu), for deciding which strings to return from a `browser.i18n.getMessage()` call.

I'm not sure how we can share the UI language with the website, since the Accept-Languages header will reflect the user's content choice, not their browser's UI language. (Is the UI language in the user agent field, maybe?) Need to figure that out, so that the addon and server always agree. (Or should we respect the user's content language choice for the website, and allow the addon strings and the website strings for the same feature to potentially disagree?)

I would appreciate any thoughts localization folks might have on this topic (@stasm? @flodolo?).

Refs #2294.

**Comment by flodolo** [comment]
CC also @zbraniecki because content vs UI language selection is a topic he's definitely familiar with (see also [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1325870)).

Personally I think it's OK to allow the add-on and website to use different languages. The website is doing the right thing (i.e. following the languages requested by the user, in their preferred order), the add-on is doing what it can, given the limitations (only one locale for UI).

More important, unlike Test Pilot, it would be good to:
* allow users to reach the website in a different language (permalinks)
* allow users to select a different language (language selector)

**Comment by zbraniecki** [comment]
Heh, interesting.

We're just now stabilizing the new API called LocaleService that has methods such as:

* LocaleService::GetAvailalbeLocales
* LocaleService::GetRequestedLocales
* LocaleService::GetAppLocalesAsBCP47

I believe you'll want the last one, but I can't commit to stability of that API just yet.

Overall, on the API level, we'll want to keep user requested locales for content separate from locales used for UI, although in Preferences most users will select those two together.

So, long story short. I have a new API, that is not stable, that gives you UI locales.

**Comment by zbraniecki** [comment]
Also, we would like to expose the new localization API for Firefox for WebExtensions as we believe it to be substantially superior to the Chrome i18n API.
That doesn't affect i18n API (you want it at least for compatibility), but wanted to raise it so that WebExtensions team knows that there's a proposal for a WebExtensions l10n API coming up that will be different from Chrome i18n API and we'll want to use it at least for in-house developed extensions.

**Comment by 6a68** [comment]
@zbraniecki @flodolo Thanks very much for the feedback. We'll probably wait to make code changes until product weighs in here, but it's good to know there are options available.

**Comment by ianb** [comment]
Since the simplest solution is to use the build language for the add-on, and the content language preference for the site, it seems best to default to the simplest thing

**Comment by zbraniecki** [comment]
> Since the simplest solution is to use the build language for the add-on, and the content language preference for the site, it seems best to default to the simplest thing

@ianb - I think we're ready to have this conversation and I belielive that we should not use content languages for web extensions (I believe we should use appLanguages from `LocaleService::GetAppLocalesAsLangTags` API)

What's the best place to hold the conversation about it? Should we file an issue, bug in bugzilla, post in some newsgroup?
